src/main/kotlin/org/jitsi/nlj/rtp/bandwidthestimation/GoogleCcEstimator.kt

```diff
@@ -0,0 +1,107 @@
+
+    /* TODO: Use configuration service to set this default value. */
+    override var initBw: Bandwidth = defaultInitBw
+    /* TODO: observable which sets the components' values if we're in initial state. */
+
+    override var minBw: Bandwidth by Delegates.observable(defaultMinBw) {
+        _, _, newValue ->
+        bitrateEstimatorAbsSendTime.setMinBitrate(newValue.bps.toInt())
+        sendSideBandwidthEstimation.setMinMaxBitrate(newValue.bps.toInt(), maxBw.bps.toInt())
+    }
+
+    override var maxBw: Bandwidth by Delegates.observable(defaultMaxBw) {
+        _, _, newValue ->
+        sendSideBandwidthEstimation.setMinMaxBitrate(minBw.bps.toInt(), newValue.bps.toInt())
+    }
+
+    private val logger = parentLogger.createChildLogger(GoogleCcEstimator::class)
+
+    /**
+     * Implements the delay-based part of Google CC.
+     */
+    private val bitrateEstimatorAbsSendTime = RemoteBitrateEstimatorAbsSendTime(diagnosticContext, logger).also {
+        it.setMinBitrate(minBw.bps.toInt())
+    }
+
+    /**
+     * Implements the loss-based part of Google CC.
+     */
+    private val sendSideBandwidthEstimation = SendSideBandwidthEstimation(diagnosticContext, initBw.bps.toLong(), logger).also {
+        it.setMinMaxBitrate(minBw.bps.toInt(), maxBw.bps.toInt())
+    }
+
+    override fun doProcessPacketArrival(now: Instant, sendTime: Instant?, recvTime: Instant?, seq: Int, size: DataSize, ecn: Byte) {
+        if (sendTime != null && recvTime != null) {
+            bitrateEstimatorAbsSendTime.incomingPacketInfo(now.toEpochMilli(),
+                    recvTime.toEpochMilli(), sendTime.toEpochMilli(), size.bytes.toInt())
+        }
+        sendSideBandwidthEstimation.updateReceiverEstimate(bitrateEstimatorAbsSendTime.latestEstimate)
+        sendSideBandwidthEstimation.updateReceiverBlock(0 /* fraction_lost */, 1 /* number_of_packets */, now.toEpochMilli())
+
+        /* TODO: rate-limit how often we call updateEstimate? */
+        sendSideBandwidthEstimation.updateEstimate(now.toEpochMilli())
+        reportBandwidthEstimate(now, sendSideBandwidthEstimation.latestEstimate.bps)
+    }
+
+    override fun doProcessPacketLoss(now: Instant, sendTime: Instant?, seq: Int) {
+        sendSideBandwidthEstimation.updateReceiverBlock(256 /* fraction_lost */, 1 /* number_of_packets */, now.toEpochMilli())
+        sendSideBandwidthEstimation.updateEstimate(now.toEpochMilli())
+        reportBandwidthEstimate(now, sendSideBandwidthEstimation.latestEstimate.bps)
+    }
+
+    override fun doRttUpdate(now: Instant, newRtt: Duration) {
+        bitrateEstimatorAbsSendTime.onRttUpdate(now.toEpochMilli(), newRtt.toMillis())
+        sendSideBandwidthEstimation.onRttUpdate(newRtt.toNanos() / 1.0e9)
+    }
+
+    override fun getCurrentBw(now: Instant): Bandwidth {
+        return sendSideBandwidthEstimation.latestEstimate.bps
+    }
+
+    override fun getStats(now: Instant): NodeStatsBlock = NodeStatsBlock("GoogleCcEstimator").apply {
```

Comment:
generally i think we've insulated these sorts of classes from  the `NodeStatsBlock` type (and made that a detail of the stats reporting elsewhere) so they've just had their own statistics object that can be converted into whatever transport type made sense by the code gathering stats.

Reply:
The thing is I wanted a generic stats object, so the abstract `BandwidthEstimator` class didn't impose too many assumptions on what stats the estimator returned.

 I agree that `NodeStatsBlock` as such is a bit semantically odd, but the semantics I wanted for these stats seemed identical to it.

Reply:
ah, i see.  what about defining something like:
```
class BandwidthEstimatorStats(
  // stats we see as common across all estimators
  val commonStatOne: Double,
  val commonStatTwo: Double,
  // sub classes can add arbitrary stat data here
  val otherStats: Map<String, Any>
)
```
?

might be able to leverage a [map property delegate](https://kotlinlang.org/docs/reference/delegated-properties.html#storing-properties-in-a-map) in the subclass to make this easier as well